drivers/i2c/i2c_shell.c

```diff
@@ -103,65 +103,108 @@ static int cmd_i2c_recover(const struct shell *shell,
 	return 0;
 }
 
-/* i2c write <device> <dev_addr> <reg_addr> [<byte1>, ...] */
-static int cmd_i2c_write(const struct shell *shell, size_t argc, char **argv)
+static int i2c_write_from_buffer(const struct shell *shell,
+		char *s_dev_name, char *s_dev_addr, char *s_reg_addr,
+		char **data, uint8_t data_length)
 {
-	uint8_t buf[MAX_I2C_BYTES];
+	/* This buffer must preserve 4 bytes for register address, as it is
+	   filled using put_be32 function and we don't want to lower available
+	   space when using 1 byte address. */
+	uint8_t buf[MAX_I2C_BYTES + 3];
 	const struct device *dev;
-	int num_bytes;
+	int reg_addr_bytes;
 	int reg_addr;
 	int dev_addr;
+	int ret;
 	int i;
 
-	dev = device_get_binding(argv[1]);
+	dev = device_get_binding(s_dev_name);
 	if (!dev) {
-		shell_error(shell, "I2C: Device driver %s not found.", argv[1]);
+		shell_error(shell, "I2C: Device driver %s not found.",
+			    s_dev_name);
 		return -ENODEV;
 	}
 
-	dev_addr = strtol(argv[2], NULL, 16);
-	reg_addr = strtol(argv[3], NULL, 16);
-	num_bytes = argc - 4;
-	if (num_bytes < 0) {
-		return 0;
-	}
-	if (num_bytes > MAX_I2C_BYTES) {
-		num_bytes = MAX_I2C_BYTES;
+	dev_addr = strtol(s_dev_addr, NULL, 16);
+	reg_addr = strtol(s_reg_addr, NULL, 16);
+
+	reg_addr_bytes = MIN(4, (strlen(s_reg_addr) + 1) / 2);
+	sys_put_be32(reg_addr, buf);
+
+	if (data_length + reg_addr_bytes > MAX_I2C_BYTES) {
+		data_length = MAX_I2C_BYTES - reg_addr_bytes;
+		shell_info(shell, "Too many bytes provided, limit is %d",
+			   MAX_I2C_BYTES - reg_addr_bytes);
 	}
-	for (i = 0; i < num_bytes; i++) {
-		buf[i] = (uint8_t)strtol(argv[4 + i], NULL, 16);
+
+	for (i = 0; i < data_length; i++) {
+		buf[4 + i] = (uint8_t)strtol(data[i], NULL, 16);
 	}
 
-	if (i2c_burst_write(dev, dev_addr, reg_addr, buf, num_bytes) < 0) {
-		shell_error(shell, "Failed to write to device: %s", argv[1]);
+	ret = i2c_write(dev, buf + 4 - reg_addr_bytes,
+			reg_addr_bytes + data_length, dev_addr);
+	if (ret < 0) {
+		shell_error(shell, "Failed to read from device: %s", s_dev_addr);
 		return -EIO;
 	}
 
 	return 0;
 }
 
+/* i2c write <device> <dev_addr> <reg_addr> [<byte1>, ...] */
+static int cmd_i2c_write(const struct shell *shell, size_t argc, char **argv)
+{
+	int ret;
+
+	ret = i2c_write_from_buffer(shell, argv[1], argv[2], argv[3],
+				    &argv[4], argc - 4);
+
+	return ret;
+}
+
 /* i2c write_byte <device> <dev_addr> <reg_addr> <value> */
 static int cmd_i2c_write_byte(const struct shell *shell,
 			      size_t argc, char **argv)
+{
+	int ret;
+
+	ret = i2c_write_from_buffer(shell, argv[1], argv[2], argv[3],
+				    &argv[4], 1);
+
+	return ret;
+}
+
+static int i2c_read_to_buffer(const struct shell *shell,
+			      char *s_dev_name,
+			      char *s_dev_addr, char *s_reg_addr,
+			      uint8_t *buf, uint8_t buf_length)
 {
 	const struct device *dev;
+	uint8_t reg_addr_buf[4];
+	int reg_addr_bytes;
 	int reg_addr;
 	int dev_addr;
-	int out_byte;
+	int ret;
```

Comment:
This seems unused at the moment?

Reply:
I am assigning the result of function to check it in the if statement. It's only used in if, but the command call is too long to wrap it directly with if